### PR TITLE
Disable interface coalescing

### LINKED ISSUE #853: Interface state tests fail when transaction queue is disabled
This is a test run with the interface state coalescing disabled:

https://gist.githubusercontent.com/pinterest-ios/f9d888dad32cb905d560a67b03f45987/raw/6de6a9b4ede5484a7d703d698678360abb6350a3/log.txt

We need to make the coalescing opt-in and also address these issues. cc @appleguy

### PR DESCRIPTION
**Pull request by wsdwsd0829**
[Fix #853] This trying to disable coalescing feature to avoid coalescing. 
It the DisplayNode change is to match behavior exactly before coalescing. Hopefully we can avoid reverting for ease of future work. 

Sorry about the troubles it causes.  I will try to fix the reported issues in #788 and give it another check. 
One potential bug fixed by #847 is one example we found recently (there maybe other issues). It's ok to submit that also since accessing pendingInterfaceState is same as self.interfaceState when the feature is disabled.

### CODE REVIEW

File: Source/ASDisplayNode.mm [hunk 1]
```diff
@@ -2898,27 +2898,28 @@ - (void)didExitHierarchy
...
-      __instanceLock__.lock();
-      unsigned isStillInHierarchy = _flags.isInHierarchy;
-      BOOL isVisible = ASInterfaceStateIncludesVisible(_pendingInterfaceState);
-      ASInterfaceState newState = (_pendingInterfaceState & ~ASInterfaceStateVisible);
-      __instanceLock__.unlock();
-
-      if (!isStillInHierarchy && isVisible) {
-        if (![self supportsRangeManagedInterfaceState]) {
-          newState = ASInterfaceStateNone;
+  if (![self supportsRangeManagedInterfaceState]) {
```
Comment (appleguy) [inline Source/ASDisplayNode.mm]: This change is equivalent to the change in the revert, when looking at specifically the disabled codepath flow - by doing this, we ensure that there are no changes to the codepath when in the disabled state.

https://github.com/TextureGroup/Texture/pull/861/files#diff-fb6dadff65468c53595c164828199c83

The root of the issue is that we made some changes to sanitize the handling of non-range managed nodes (e.g. top level nodes / outside of cells) so they behaved more like inside cells. We now have Texture tests running on our internal build system and so we would have caught this at the time, but alas we didn’t notice this broke the tests in the disabled case.

Definitely an oversight on our part, which I think won’t happen again considering the build system and ASConfiguration systems in place. Fortunately I think we are on the path to fixing some long-standing issues where visible state will thrash during UINavigationController transitions, which will be a big help for logging accuracy.

File: Source/ASDisplayNode.mm [hunk 2]
```diff
@@ -2898,27 +2898,48 @@ - (void)didExitHierarchy
...
-      BOOL isVisible = ASInterfaceStateIncludesVisible(_pendingInterfaceState);
-      ASInterfaceState newState = (_pendingInterfaceState & ~ASInterfaceStateVisible);
-      __instanceLock__.unlock();
-
-      if (!isStillInHierarchy && isVisible) {
-        if (![self supportsRangeManagedInterfaceState]) {
-          newState = ASInterfaceStateNone;
-        }
-        self.interfaceState = newState;
+  if (!ASCATransactionQueue.sharedQueue.enabled) {
```
Comment (wsdwsd0829) [inline Source/ASDisplayNode.mm]: This is to add back reverted code before interface coalescing. 
An put current code in else for ease for read & future debugging.

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: Source/ASDisplayNode.mm [hunk 3]
```diff
@@ -2898,27 +2898,48 @@ - (void)didExitHierarchy
...
+          __instanceLock__.unlock();
+          if (!isInHierarchy && isVisible) {
+            self.interfaceState = newState;
+          }
+        });
       }
-    };
+    }
+  } else {
+    // TODO(wsdwsd0829): rework on this piece code that alter interface behavior of non-interfaceCoalescing.
```
Comment (appleguy) [inline Source/ASDisplayNode.mm]: // Rework this code to ensure that interface state behavior is not altered when ASCATransactionQueue is disabled.

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: Source/ASDisplayNode.mm [hunk 4]
```diff
@@ -58,6 +58,8 @@
   #define TIME_SCOPED(outVar)
 #endif
 
+#define ENABLE_NEW_EXIT_HIERARCHY_BEHAVIOR 0
```
Comment (Adlai-Holler) [inline Source/ASDisplayNode.mm]: Would you be willing to add a comment describing this macro?

Comment (wsdwsd0829) [reply Source/ASDisplayNode.mm]: Hi, you mean I added in code or here?
The new NEW_EXIT_HIERARCHY_BEHAVIOR is trying to merge non-rangeManaged with rangeManaged, so both range-managed and standalone nodes wait before firing their exit-visibility handlers, as UIViewController transitions now do rehosting at both start & end of animation. (that was to try to mitigate interface updating state when coalescing disabled)

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: Source/ASDisplayNode.mm (hunk 4 above)
Comment (Adlai-Holler) [reply Source/ASDisplayNode.mm]: OK that makes sense. Could you add that on line 60 so that future developers will know the intention?

Comment (wsdwsd0829) [reply Source/ASDisplayNode.mm]: Done

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by Adlai-Holler** [comment]
Thanks for the fast fIx here @appleguy and @wsdwsd0829

**Comment by wsdwsd0829** [comment]
Thanks for the quick reply.

**Pull request merged**